environment/src/main/java/com/sequenceiq/environment/environment/flow/creation/handler/freeipa/FreeIpaCreationRetrievalTask.java

```diff
@@ -19,44 +22,52 @@
+            } else if (freeIpa.get().getStatus().isDeletionInProgress() || freeIpa.get().getStatus().isSuccessfullyDeleted()) {
+                LOGGER.error("FreeIpa '{}' '{}' is getting terminated (status:'{}'), polling is cancelled.",
+                        freeIpa.get().getName(),
+                        freeIpa.get().getCrn(),
+                        freeIpa.get().getStatus());
                 throw new FreeIpaOperationFailedException("FreeIpa instance deleted under the creation process.");
-            } else if (freeIpa.getStatus().isFailed()) {
-                LOGGER.info("FreeIpa '{}' '{}' is in failed state (status:'{}'), polling is cancelled.",
-                        freeIpa.getName(),
-                        freeIpa.getCrn(),
-                        freeIpa.getStatus());
+            } else if (freeIpa.get().getStatus().isFailed()) {
+                LOGGER.error("FreeIpa '{}' '{}' is in failed state (status:'{}'), polling is cancelled.",
+                        freeIpa.get().getName(),
+                        freeIpa.get().getCrn(),
+                        freeIpa.get().getStatus());
                 throw new FreeIpaOperationFailedException(String.format("FreeIpa creation failed. Status: '%s' statusReason: '%s'",
-                        freeIpa.getStatus(), freeIpa.getStatusReason()));
-            } else if (freeIpa.getStatus().isAvailable()) {
+                        freeIpa.get().getStatus(), freeIpa.get().getStatusReason()));
+            } else if (freeIpa.get().getStatus().isAvailable()) {
                 return true;
             }
-        } catch (FreeIpaOperationFailedException e) {
-            throw e;
         } catch (Exception e) {
-            throw new FreeIpaOperationFailedException(e.getMessage(), e);
+            throw new FreeIpaOperationFailedException("FreeIpa creation operation failed. " + e.getMessage(), e);
         }
         return false;
     }
 
     @Override
     public void handleTimeout(FreeIpaPollerObject freeIpaPollerObject) {
         try {
-            DescribeFreeIpaResponse freeIpa = freeIpaPollerObject.getFreeIpaV1Endpoint().describe(freeIpaPollerObject.getEnvironmentCrn());
+            Optional<DescribeFreeIpaResponse> freeIpa = freeIpaService.describe(freeIpaPollerObject.getEnvironmentCrn());
+            if (freeIpa.isEmpty()) {
+                throw new FreeIpaOperationFailedException("FreeIpa cluster was not found for environment: "
+                        + freeIpaPollerObject.getEnvironmentCrn());
+            }
             throw new FreeIpaOperationFailedException(String.format("Polling operation timed out, FreeIpa creation failed. FreeIpa status: '%s' "
-                    + "statusReason: '%s'", freeIpa.getStatus(), freeIpa.getStatusReason()));
+                    + "statusReason: '%s'", freeIpa.get().getStatus(), freeIpa.get().getStatusReason()));
```

Comment:
same as above, get the optional value only once... but since this is only 2 calls, you may omit this comment.